### PR TITLE
Setting current color to black if canvas is not rendered in document

### LINKED ISSUE #10601: CanvasRenderingContext2D currentColor support is broken
Blocks #10597.

At least for the not-rendered case. `style.GetPropertyValue(DOMString::from("color"))` returns the empty string, and parsing that fails.

* Spec: around <https://html.spec.whatwg.org/multipage/scripting.html#dom-context-2d-height>.
* Code: `components/script/dom/canvasrenderingcontext2d.rs`.
* Test: `2dcontext/fill-and-stroke-styles/2d.fillStyle.parse.current.removed.html`.

### PR DESCRIPTION
**Pull request by craftytrickster**
Fixes #10601

The change seems deceptively easy, I hope I am not missing anything...

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10651)

### CODE REVIEW

**Comment by emilio** [comment]
The change looks good, but the 2d context spec defines being rendered quite precisely, and it's not what we're doing here:

> An element is being rendered if it has any associated CSS layout boxes, SVG layout boxes, or some equivalent in other styling languages.

That is, if the node has an associated layout fragment.

I think testing for presence in the document and the computed value of the `display` property (checking it's not `none`) could be enough, over all knowing we already use `GetComputedStyle` (you can make that change and I'll trigger a try run).

I'm not sure if there's another way for an element not to generate a layout box (and if that's well specified enough). If the condition is more complex, we might need to query layout, but...

I'll check the CSS spec and other browsers' implementation, meanwhile that change is probably saner :)

-S-awaiting-review +S-needs-code-changes 

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks broke.

---

*[components/script/dom/canvasrenderingcontext2d.rs, line 497 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10651#-KFUx5g5tz2SnGjlsCBX:-KFUx5g5tz2SnGjlsCBY:-1472283869) ([raw file](https://github.com/servo/servo/blob/6614076a58dfa46ce263b91f48443bd4f1c6c979/components/script/dom/canvasrenderingcontext2d.rs#L497)):*
nit: `} else {`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10651#-:-KFUxFE7MReQG4eCDXJk:-563044224)*

**Comment by craftytrickster** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/canvasrenderingcontext2d.rs, line 497 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10651#-KFUx5g5tz2SnGjlsCBX:-KFV4fRw3FOBhEilLxU8:-499146009) ([raw file](https://github.com/servo/servo/blob/6614076a58dfa46ce263b91f48443bd4f1c6c979/components/script/dom/canvasrenderingcontext2d.rs#L497)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10651)*

**Comment by craftytrickster** [comment]
When I have time later I will try to add the computed value of the display as well. 

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10651#-:-KFV5WpMGIAMt9i774nx:-883423236)*

**Comment by jdm** [comment]
`display: none` is one way to make an element not have a layout box.

**Comment by emilio** [comment]
@jdm: Yeah my concern is if it's the only one or not :)

**Comment by dzbarsky** [comment]
I remember looking at this when I was implementing, and the spec doesn't match what existing browsers do. Don't remember the exact details, but please tread carefully! (and file a spec bug after investigation, like I should have done before)

**Comment by craftytrickster** [comment]
Assuming that the `element_not_rendered` implementation is correct, should the logic to get the boolean value be placed somewhere else? If it were placed as an impl method somewhere (not sure which place would be the most appropriate), then the logic could be reused.

**Comment by craftytrickster** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/canvasrenderingcontext2d.rs, line 498 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10651#-KF_o4yCHg5tgMeJxpf2:-KF_o4yCHg5tgMeJxpf3:1994195981) ([raw file](https://github.com/servo/servo/blob/7d08beee57ae2448ce4b108d1c95458f569ec541/components/script/dom/canvasrenderingcontext2d.rs#L498)):*
If not_in_doc is true, then there is no need to perform the display_is_none check. I can make the change once I know if this logic will be extracted into its own method or not.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10651)*

**Comment by emilio** [comment]
I don't think it's actually needed to extract it in its own method, probably using short-circuiting or would suffice :)

Let's check if this matches what the 2dcontext test suite expects :)

@bors-servo: try

**Comment by craftytrickster** [comment]
The test says it fails, yet if I run the test, it do see the canvas getting set to black. Is this an implementation error or does the test itself need to be changed?

**Comment by KiChjang** [comment]
@craftytrickster The test says it PASSES, doesn' it? It's just expected to FAIL instead.

**Comment by craftytrickster** [comment]
@KiChjang I see that too, but it says that the expected output is a black rectangle. Prior to the change the output was a red rectangle, but now it is indeed black, so why would it be considered a failure? Is it possible it was just set to expect a failure to pass CI?

**Comment by KiChjang** [comment]
@craftytrickster Nonononono, the test is currently passing, but our test suite expects it to fail. What you need to do in this case is to change the test expectations, i.e. remove [this file](https://github.com/servo/servo/blob/master/tests/wpt/metadata/2dcontext/fill-and-stroke-styles/2d.fillStyle.parse.current.removed.html.ini).

**Comment by craftytrickster** [comment]
@KiChjang thanks for the pointer on the .ini file, I have removed it as suggested.

**Comment by Ms2ger** [comment]
Right now, we only have a test for the `is_in_doc` check, so I'd like you to add a test where:

1. the canvas element is in the document but *it* is `display:none`
1. the canvas element is in the document but *its parent* is `display:none`

and check what other browsers do there.

**Comment by craftytrickster** [comment]
https://github.com/whatwg/html/issues/1099 affects this PR.

File: components/script/dom/canvasrenderingcontext2d.rs [hunk 1]
```diff
@@ -486,11 +486,24 @@ impl CanvasRenderingContext2D {
...
+
                     let style = window.GetComputedStyle(&*self.canvas.upcast(), None);
-                    self.parse_color(&style.GetPropertyValue(DOMString::from("color")))
+
+                    let element_not_rendered =
+                        !self.canvas.upcast::<Node>().is_in_doc() ||
+                        style.GetPropertyValue(DOMString::from("display")) == "none";
+
+                    if element_not_rendered {
+                        self.parse_color("black")
```
Comment (Ms2ger) [inline components/script/dom/canvasrenderingcontext2d.rs]: Let's make this

```rust
Ok(RGBA {
    red: 0.0,
    green: 0.0,
    blue: 0.0,
    alpha: 0.0,
})
```

Comment (craftytrickster) [inline components/script/dom/canvasrenderingcontext2d.rs]: Done

**Pull request merged**